ext/web/hr_timer_lock.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+#[cfg(target_os = "windows")]
+mod windows {
+  use std::marker::PhantomData;
+  use std::sync::atomic::AtomicU32;
+
+  pub(crate) struct HrTimerLock {
+    pub(super) _unconstructable: PhantomData<()>,
+  }
+
+  /// Decrease the reference count of the HR timer on drop.
+  impl Drop for HrTimerLock {
+    fn drop(&mut self) {
+      dec_ref();
+    }
+  }
+
+  /// Maintains the HR timer refcount. Times should not be nested more than 5 deep so this is
+  /// more than sufficient.
+  static TIMER_REFCOUNT: AtomicU32 = AtomicU32::new(0);
+
+  pub(super) fn inc_ref() {
+    let old = TIMER_REFCOUNT.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    // Overflow/underflow sanity check in debug mode
+    debug_assert!(old != u32::MAX);
+    if old == 0 {
+      lock_hr();
+    }
+  }
+
+  fn dec_ref() {
+    let old = TIMER_REFCOUNT.fetch_sub(1, std::sync::atomic::Ordering::SeqCst);
+    // Overflow/underflow sanity check in debug mode
+    debug_assert!(old != 0);
+    if old == 1 {
+      unlock_hr();
```

Comment:
Someone looking at this code might think there's a race condition here with multiple threads (with workers), but winapi handles multiple calls to timeBeginPeriod and timeEndPeriod. So for example (I tested and verified this is the case):

```
// resolution: 15

timeBeginPeriod(1); // thread 1
// resolution: 1

timeBeginPeriod(1); // thread 2
// resolution: 1

timeEndPeriod(1); // thread 1
// resolution: 1

timeEndPeriod(1); // thread 2
// resolution: 15
```

Maybe it would be good to add a comment somewhere explaining that winapi tracks multiple calls to timeBeginPeriod.

Reply:
In this case I decided to do the refcount ourselves like Chrome does (manage our own lock and set the OS period only when we have a refcount >= 1). I am not sure if other platforms might need a HR timer lock so this allows us to work with them if the time comes.

Reply:
What I meant is I believe the following scenario can occur:

```
[Worker]: Increment atomic (1)
[Worker]: Lock hr
[Worker]: Decrement atomic (0)
[Main]: Increment atomic (1)
[Main]: Lock hr
[Worker]: Unlock hr
```

So we're still relying on the winapi behaviour in some very rare scenarios, which is fine. The new comment is good.